src/plonkish/backend/halo2.rs

```diff
@@ -671,39 +672,65 @@ fn assign_witness<F: PrimeField + From<u64>>(
     }
 }
 
-pub trait PlonkishHalo2<F: PrimeField, W, WG: Halo2WitnessGenerator<F, W>> {
+#[allow(private_bounds)]
+pub trait Halo2Provable<W, WG: Halo2WitnessGenerator<Fr, W>>: Halo2Compilable<W, WG> {
     /// Create a Halo2 prover
     ///
     /// ### Arguments
-    /// * `rng` - random number generator
+    /// * `params_path` - path to trusted setup file
     ///
     /// ### Returns
     /// * a Halo2 prover
-    fn create_halo2_prover(&mut self, rng: BlockRng<DummyRng>) -> Halo2Prover<F, W, WG>;
+    fn create_halo2_prover(&mut self, params_path: &str) -> Halo2Prover<Fr, W, WG> {
+        let (circuit, compiled, k) = self.halo2_compile();
+        let mut params_fs = File::open(params_path).expect("couldn't load params");
+        let mut params = ParamsKZG::<Bn256>::read(&mut params_fs).expect("Failed to read params");
+        if params.k() < k {
+            panic!(
+                "The provided trusted setup size {} ({params_path}) does not satisfy the circuit size {k}",
+                params.k(),
+            );
+        }
+        if params.k() > k {
+            params.downsize(k);
+        }
+        Halo2Prover::new(create_setup(compiled, params), circuit)
+    }
+
+    /// Create a Halo2 test prover. ⚠️ Not for production use! ⚠️
+    /// This prover uses a dummy RNG that outputs fixed values.
+    ///
+    /// ### Returns
+    /// * a test Halo2 prover
+    fn create_test_prover(&mut self) -> Halo2Prover<Fr, W, WG> {
+        let (circuit, compiled, k) = self.halo2_compile();
+
+        let params = ParamsKZG::<Bn256>::setup::<BlockRng<DummyRng>>(k, BlockRng::new(DummyRng {}));
+
+        Halo2Prover::new(create_setup(compiled, params), circuit)
+    }
 }
 
-impl<TG: TraceGenerator<Fr> + Default> PlonkishHalo2<Fr, Assignments<Fr>, ChiquitoHalo2<Fr>>
+impl<W, WG: Halo2WitnessGenerator<Fr, W>, T: Halo2Compilable<W, WG>> Halo2Provable<W, WG> for T {}
```

Comment:
@alxkzmn I think there has been a misunderstanding again.

Although this blanket trait could be useful, I don't say to remove it.

I am saying to do:

```
impl<T: Halo2Compilable>  Halo2Provable for T {
    fn create_halo2_prover( ....
}
```

that way you don't need `#[allow(private_bounds)]`

Reply:
With the trait `Halo2Provable` without default implementation for `create_halo2_prover`